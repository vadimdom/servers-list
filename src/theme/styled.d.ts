import 'styled-components';

import { Theme } from './defaultTheme';

declare module 'styled-components' {
  export interface DefaultTheme extends Theme {}
}
