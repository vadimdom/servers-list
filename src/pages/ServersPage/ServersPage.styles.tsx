import { styled } from 'styled-components';
import { DEVICE_SIZES } from '../../constants';

export const PageWrapper = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  padding: 70px 30px 50px 30px;
  min-height: 100vh;
`;

export const PageDescription = styled.p`
  font-size: 20px;
  font-weight: bold;
  text-align: center;
  color: ${({ theme }) => theme.color.darkBlue};
  background: rgba(255, 255, 255, 0.5);
  border-radius: 8px;
  padding: 10px;

  @media (max-width: ${DEVICE_SIZES.tablet}) {
    font-size: 16px;
  }
`;

export const ServersList = styled.div`
  display: flex;
  flex-direction: column;
  padding: 20px;
  background: rgba(255, 255, 255, 0.6);
  color: ${({ theme }) => theme.color.darkBlue};
  border-radius: 8px;
  width: 100%;
  max-width: 600px;

  @media (max-width: ${DEVICE_SIZES.tablet}) {
    max-width: 80%;
    padding: 10px;
  }
`;

export const Servers = styled.div`
  display: flex;
  flex-direction: column;
  padding-top: 10px;
  gap: 8px;
`;

export const ServersListHeader = styled.div`
  display: flex;
  justify-content: space-between;
`;

export const ListHeader = styled.div`
  font-size: 18px;
  font-weight: bold;
  display: flex;
  align-items: center;
  gap: 8px;
  cursor: pointer;

  @media (max-width: ${DEVICE_SIZES.tablet}) {
    font-size: 14px;
  }
`;
