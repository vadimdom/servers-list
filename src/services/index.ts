export { login } from './auth';
export { getServers } from './servers';
