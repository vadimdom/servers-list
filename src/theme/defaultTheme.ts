import { COLORS } from '../constants';

export const defaultTheme = {
  color: { ...COLORS },
};

export type Theme = typeof defaultTheme;
