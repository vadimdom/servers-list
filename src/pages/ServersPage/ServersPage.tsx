import { useEffect, useState } from 'react';

import { getServers } from '../../services';
import { Loader, ServerItem, SortMarker } from '../../components';
import { ServerType, SortRuleType } from '../../types';
import {
  PageWrapper,
  PageDescription,
  ServersList,
  Servers,
  ServersListHeader,
  ListHeader,
} from './ServersPage.styles';
import { ascendingSort, descendingSort } from '../../helpers';

export const ServersPage = () => {
  const [servers, setServers] = useState<ServerType[]>([]);
  const [isLoaded, setIsLoaded] = useState<boolean>(false);
  const [nameSortRule, setNameSortRule] = useState<SortRuleType>(null);
  const [distanceSortRule, setDistanceSortRule] = useState<SortRuleType>(null);

  useEffect(() => {
    const fetchServers = async () => {
      const result = await getServers();

      setIsLoaded(true);
      setServers(result);
    };

    fetchServers();
  }, []);

  const sortByName = () => {
    const sortedServers = [...servers].sort((a, b) =>
      nameSortRule === 'asc' ? descendingSort(a.name, b.name) : ascendingSort(a.name, b.name)
    );
    setNameSortRule((prevNameSortRule) => (prevNameSortRule === 'asc' ? 'desc' : 'asc'));
    setDistanceSortRule(null);
    setServers(sortedServers);
  };

  const sortByDistance = () => {
    const sortedServers = [...servers].sort((a, b) =>
      distanceSortRule === 'asc' ? descendingSort(a.distance, b.distance) : ascendingSort(a.distance, b.distance)
    );
    setDistanceSortRule((prevNameSortRule) => (prevNameSortRule === 'asc' ? 'desc' : 'asc'));
    setNameSortRule(null);
    setServers(sortedServers);
  };

  return (
    <PageWrapper data-testid="servers-page">
      <PageDescription>Here is the list of all available servers:</PageDescription>
      {isLoaded ? (
        <ServersList>
          {servers.length ? (
            <>
              <ServersListHeader>
                <ListHeader data-cy="sort-by-name" data-testid="sort-by-name" onClick={sortByName}>
                  <span>Name</span>
                  <SortMarker rule={nameSortRule} />
                </ListHeader>
                <ListHeader data-cy="sort-by-distance" data-testid="sort-by-distance" onClick={sortByDistance}>
                  <span>Distance</span>
                  <SortMarker rule={distanceSortRule} />
                </ListHeader>
              </ServersListHeader>
              <Servers data-cy="servers-list" data-testid="servers-list">
                {servers.map((server) => (
                  <ServerItem key={`${server.name}-${server.distance}`} server={server} />
                ))}
              </Servers>
            </>
          ) : (
            <span>No available servers</span>
          )}
        </ServersList>
      ) : (
        <Loader data-testid="loader" $sizePX={100} />
      )}
    </PageWrapper>
  );
};
