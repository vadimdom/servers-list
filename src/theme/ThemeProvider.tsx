import { ReactNode } from 'react';
import { ThemeProvider as StyledThemeProvider } from 'styled-components';

import { defaultTheme } from './defaultTheme';

type ThemeProviderProps = {
  children: ReactNode;
};
export const ThemeProvider = ({ children }: ThemeProviderProps) => {
  return <StyledThemeProvider theme={defaultTheme}>{children}</StyledThemeProvider>;
};
