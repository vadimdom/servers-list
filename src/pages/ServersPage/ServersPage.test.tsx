import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';

import { ServersPage } from './ServersPage';
import { ThemeProvider } from '../../theme';
import { getServers } from '../../services';

jest.mock('../../services');

const mockedUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedUseNavigate,
}));

const renderServersPage = () => {
  render(
    <MemoryRouter>
      <ThemeProvider>
        <ServersPage />
      </ThemeProvider>
    </MemoryRouter>
  );
};

const localStorageMock = (() => {
  return {
    setItem: () => null,
  };
})();

Object.defineProperty(window, 'localStorage', {
  value: localStorageMock,
});
const LIST_OF_SERVERS = [
  { name: 'First', distance: 711 },
  { name: 'Third', distance: 123 },
  { name: 'Second', distance: 1213 },
  { name: 'Second', distance: 123 },
];

describe('<ServersPage />', () => {
  beforeEach(async () => {
    (getServers as jest.Mock).mockReturnValue(LIST_OF_SERVERS);
  });

  test('renders loader, description and no servers message', async () => {
    (getServers as jest.Mock).mockReturnValue([]);
    renderServersPage();

    expect(screen.getByTestId('loader')).toBeInTheDocument();
    expect(screen.getByTestId('servers-page')).toHaveTextContent(/list of all available servers/i);
    await waitFor(() => expect(screen.getByTestId('servers-page')).toHaveTextContent(/No available servers/i));
  });
  test('renders correct amount of servers', async () => {
    (getServers as jest.Mock).mockReturnValue([LIST_OF_SERVERS[0]]);
    renderServersPage();

    const serversList = await screen.findByTestId('servers-list');
    expect(serversList.children).toHaveLength(1);
  });
  test('should sort by server name', async () => {
    renderServersPage();

    const sortByNameButton = await screen.findByTestId('sort-by-name');

    fireEvent.click(sortByNameButton);

    const serversList = await screen.findByTestId('servers-list');

    expect(serversList.children).toHaveLength(LIST_OF_SERVERS.length);
    expect(serversList.children[0]).toHaveTextContent(LIST_OF_SERVERS[0].name);
    expect(serversList.children[1]).toHaveTextContent(LIST_OF_SERVERS[2].name);
    expect(serversList.children[2]).toHaveTextContent(LIST_OF_SERVERS[3].name);
    expect(serversList.children[3]).toHaveTextContent(LIST_OF_SERVERS[1].name);

    fireEvent.click(sortByNameButton);
    expect(serversList.children[0]).toHaveTextContent(LIST_OF_SERVERS[1].name);
    expect(serversList.children[1]).toHaveTextContent(LIST_OF_SERVERS[2].name);
    expect(serversList.children[2]).toHaveTextContent(LIST_OF_SERVERS[3].name);
    expect(serversList.children[3]).toHaveTextContent(LIST_OF_SERVERS[0].name);
  });
  test('should sort by server distance', async () => {
    renderServersPage();

    const sortByDistanceButton = await screen.findByTestId('sort-by-distance');

    fireEvent.click(sortByDistanceButton);

    const serversList = await screen.findByTestId('servers-list');

    expect(serversList.children).toHaveLength(LIST_OF_SERVERS.length);
    expect(serversList.children[0]).toHaveTextContent(LIST_OF_SERVERS[1].name);
    expect(serversList.children[1]).toHaveTextContent(LIST_OF_SERVERS[3].name);
    expect(serversList.children[2]).toHaveTextContent(LIST_OF_SERVERS[0].name);
    expect(serversList.children[3]).toHaveTextContent(LIST_OF_SERVERS[2].name);

    fireEvent.click(sortByDistanceButton);
    expect(serversList.children[0]).toHaveTextContent(LIST_OF_SERVERS[2].name);
    expect(serversList.children[1]).toHaveTextContent(LIST_OF_SERVERS[0].name);
    expect(serversList.children[2]).toHaveTextContent(LIST_OF_SERVERS[1].name);
    expect(serversList.children[3]).toHaveTextContent(LIST_OF_SERVERS[3].name);
  });
});
