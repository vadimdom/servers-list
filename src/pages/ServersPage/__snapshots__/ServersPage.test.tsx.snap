// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ServersPage /> renders loader, description and no servers message 1`] = `
.c3 {
  display: flex;
  gap: 6px;
  align-items: center;
  font-size: 12px;
  font-weight: bold;
}

.c5 {
  width: 20px;
  height: 20px;
  cursor: pointer;
  border-radius: 50%;
  opacity: 0.5;
  border: 2px solid #000000;
}

.c4 {
  display: flex;
  gap: 2px;
  align-items: center;
}

.c0 {
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 70px 30px 50px 30px;
  min-height: calc(100vh - 120px);
}

.c1 {
  font-size: 20px;
  font-weight: bold;
  text-align: center;
  color: #234d73;
  background: rgba(255, 255, 255, 0.5);
  border-radius: 8px;
  padding: 10px;
}

.c2 {
  display: flex;
  flex-direction: column;
  gap: 8px;
  padding: 20px;
  background: rgba(255, 255, 255, 0.6);
  color: #234d73;
  border-radius: 8px;
  width: 100%;
  max-width: 600px;
}

@media (max-width: 768px) {
  .c1 {
    font-size: 16px;
  }
}

@media (max-width: 768px) {
  .c2 {
    max-width: 80%;
    padding: 10px;
  }
}

<div
  class="c0"
  data-testid="servers-page"
>
  <p
    class="c1"
  >
    Here is the list of all available servers:
  </p>
  <div
    class="c2"
  >
    <div
      class="c3"
    >
      <span>
        Filter by countries:
      </span>
      <div
        class="c4"
        data-testid="country-flags"
      >
        <img
          class="c5"
          src="uk.svg"
        />
        <img
          class="c5"
          src="us.svg"
        />
        <img
          class="c5"
          src="de.svg"
        />
        <img
          class="c5"
          src="ja.svg"
        />
      </div>
    </div>
    <span>
      No available servers
    </span>
  </div>
</div>
`;

exports[`<ServersPage /> should sort by server name 1`] = `
.c9 {
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 4px 20px;
  background: #ffffff;
  border-radius: 4px;
  gap: 10px;
  box-shadow: rgba(79, 192, 249, 0.24) 0px 3px 8px;
  transition: all 0.1s linear;
  font-size: 16px;
  font-weight: bold;
}

.c9:hover {
  box-shadow: rgba(79, 192, 249, 0.5) 0px 4px 15px;
  transform: translateY(1px);
  background: #ecf5fd;
}

.c11 {
  width: 30px;
  height: 30px;
}

.c10 {
  display: flex;
  align-items: center;
  gap: 10px;
}

.c12 {
  width: 20px;
  height: 20px;
}

.c13 {
  color: #ffa500;
}

.c14 {
  color: #008000;
}

.c3 {
  display: flex;
  gap: 6px;
  align-items: center;
  font-size: 12px;
  font-weight: bold;
}

.c5 {
  width: 20px;
  height: 20px;
  cursor: pointer;
  border-radius: 50%;
  opacity: 0.5;
  border: 2px solid #000000;
}

.c4 {
  display: flex;
  gap: 2px;
  align-items: center;
}

.c0 {
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 70px 30px 50px 30px;
  min-height: calc(100vh - 120px);
}

.c1 {
  font-size: 20px;
  font-weight: bold;
  text-align: center;
  color: #234d73;
  background: rgba(255, 255, 255, 0.5);
  border-radius: 8px;
  padding: 10px;
}

.c2 {
  display: flex;
  flex-direction: column;
  gap: 8px;
  padding: 20px;
  background: rgba(255, 255, 255, 0.6);
  color: #234d73;
  border-radius: 8px;
  width: 100%;
  max-width: 600px;
}

.c8 {
  display: flex;
  flex-direction: column;
  gap: 8px;
}

.c6 {
  display: flex;
  justify-content: space-between;
}

.c7 {
  font-size: 18px;
  font-weight: bold;
  display: flex;
  align-items: center;
  gap: 8px;
  cursor: pointer;
}

@media (max-width: 768px) {
  .c9 {
    font-size: 12px;
  }
}

@media (max-width: 768px) {
  .c11 {
    display: none;
  }
}

@media (max-width: 768px) {
  .c1 {
    font-size: 16px;
  }
}

@media (max-width: 768px) {
  .c2 {
    max-width: 80%;
    padding: 10px;
  }
}

@media (max-width: 768px) {
  .c7 {
    font-size: 14px;
  }
}

<div
  class="c0"
  data-testid="servers-page"
>
  <p
    class="c1"
  >
    Here is the list of all available servers:
  </p>
  <div
    class="c2"
  >
    <div
      class="c3"
    >
      <span>
        Filter by countries:
      </span>
      <div
        class="c4"
        data-testid="country-flags"
      >
        <img
          class="c5"
          src="uk.svg"
        />
        <img
          class="c5"
          src="us.svg"
        />
        <img
          class="c5"
          src="de.svg"
        />
        <img
          class="c5"
          src="ja.svg"
        />
      </div>
    </div>
    <div
      class="c6"
    >
      <div
        class="c7"
        data-cy="sort-by-name"
        data-testid="sort-by-name"
      >
        <span>
          Name
        </span>
      </div>
      <div
        class="c7"
        data-cy="sort-by-distance"
        data-testid="sort-by-distance"
      >
        <span>
          Distance
        </span>
      </div>
    </div>
    <div
      class="c8"
      data-cy="servers-list"
      data-testid="servers-list"
    >
      <div
        class="c9"
        data-cy="server-item"
      >
        <div
          class="c10"
        >
          <svg
            class="c11"
          >
            server.svg
          </svg>
          <img
            class="c12"
            src="de.svg"
          />
          <span
            data-cy="server-name"
          >
            Germany
          </span>
        </div>
        <div
          class="c13"
          data-cy="server-distance"
        >
          711
        </div>
      </div>
      <div
        class="c9"
        data-cy="server-item"
      >
        <div
          class="c10"
        >
          <svg
            class="c11"
          >
            server.svg
          </svg>
          <img
            class="c12"
            src=""
          />
          <span
            data-cy="server-name"
          >
            Third
          </span>
        </div>
        <div
          class="c14"
          data-cy="server-distance"
        >
          123
        </div>
      </div>
      <div
        class="c9"
        data-cy="server-item"
      >
        <div
          class="c10"
        >
          <svg
            class="c11"
          >
            server.svg
          </svg>
          <img
            class="c12"
            src="ja.svg"
          />
          <span
            data-cy="server-name"
          >
            Japan
          </span>
        </div>
        <div
          class="c13"
          data-cy="server-distance"
        >
          1213
        </div>
      </div>
      <div
        class="c9"
        data-cy="server-item"
      >
        <div
          class="c10"
        >
          <svg
            class="c11"
          >
            server.svg
          </svg>
          <img
            class="c12"
            src="ja.svg"
          />
          <span
            data-cy="server-name"
          >
            Japan
          </span>
        </div>
        <div
          class="c14"
          data-cy="server-distance"
        >
          123
        </div>
      </div>
    </div>
  </div>
</div>
`;
